### PR TITLE
[feat] provide `update` to `use:enhance` callback

### LINKED ISSUE #7072: Provide the default callback to `enhance`
### Describe the problem

I've seen people try to add a loading state via enhance a couple times. They do something like this:

```
<form use:enhance={() => {
  loading = true;
  return () => {
    loading = false;
  }
}}>
```

However, they don't realize that once you return a callback from the function passed to enhance, the default behavior after a form submission no longer applies and you have to do the following yourself:

- run `invalidateAll` if the result was a success
- run `applyAction` if the current location matches or the result was a redirect/error

This logic currently happens in the [default callback](https://github.com/sveltejs/kit/blob/fdb166cb80949c2f54093a7787bcd5eb43524666/packages/kit/src/runtime/app/forms.js#L26-L40).

This means if you want to extend `enhance`, e.g. to add a loading state, you have to duplicate that logic, even if your customizations are minimal:

```diff
<form use:enhance={() => {
  loading = true;
  return ({action, result}) => {
+    if (result.type === 'success') {
+      await invalidateAll();
+    }
+
+    if (
+      location.origin + location.pathname === action.origin + action.pathname ||
+      result.type === 'redirect' ||
+      result.type === 'error'
+    ) {
+      applyAction(result);
+    }

    loading = false;
  }
}}>
```

(You could remove the check around `applyAction` and always apply it, but I don't like how that diverges you from the default `enhance` behavior, perhaps without even realizing it).

You can abstract this into a custom action that wraps `enhance` to reduce the boilerplate, but you won't get any updates that are made to the default callback behavior, e.g. the changes in https://github.com/sveltejs/kit/pull/6828.

All you wanted to do was toggle a loading state, but now you have to maintain the default `enhance` behavior as well.

Of course, there are cases where you want to customize the invalidate or `applyAction` behavior. However, I suspect those cases are more the exception than the norm.

### Describe the proposed solution

Provide an additional argument to the callback returned from enhance. Name is bikesheddable, but for the purposes of this issue I'll call it `defaultBehavior`. The implementation of this method is the same as [fallback_callback](https://github.com/sveltejs/kit/blob/fdb166cb80949c2f54093a7787bcd5eb43524666/packages/kit/src/runtime/app/forms.js#L26-L40). People who want to add some logic after a form submission _and_ preserve the default behavior could call this method instead of recreating the logic themselves. This way, if the default behavior is updated, custom actions calling this method get it for free.

The example above would then look like this:

```html
<form use:enhance={() => {
  loading = true;
  return ({action, result, defaultBehavior}) => {
    defaultBehavior();
    loading = false;
  }
}}>
```

If people want to write enhance callbacks that _always_ do the default callback behavior, then they could create their own action that calls `defaultBehavior` without having to track any updates to the behavior.

### Alternatives considered

One alternative is to keep the status quo and require the user to duplicate the default behavior if desired.

Another alternative is to call `fallback_callback` automatically after the returned callback runs, but also pass a `cancel` arg to the returned callback that would prevent that behavior if desired. 

```html
<form use:enhance={() => {
  loading = true;
  return ({action, result, cancel}) => {
    // invalidate & applyAction called automatically, unless cancel is called
    loading = false;
  }
}}>
```

This makes the default case simpler. However, this has a few downsides:

- it's a breaking change for people returning a callback that didn't want `invalidateAll` and `applyAction` called
- because we need to run the provided callback to find out if `cancel` was called, there isn't a good way to run code after invalidation & applying the action has completed. Supplying `defaultBehavior` gives the user control over when invalidation happens.

### Importance

would make my life easier

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by dummdidumm**
Closes #7072

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. All changesets should be `patch` until SvelteKit 1.0

### CODE REVIEW

**Comment by Rich-Harris** [comment]
What about `update` instead of `defaultBehavior`?

File: documentation/docs/06-form-actions.md [hunk 1]
```diff
@@ -257,18 +257,20 @@ Without an argument, `use:enhance` will emulate the browser-native behaviour, ju
...
 > By default the `form` property and `$page.form` is only updated for actions that are in a `+page.server.js` alongside the `+page.svelte` because in the native form submission case you would be redirected to the page the action is on
 
-To customise the behaviour, you can provide a function that runs immediately before the form is submitted, and (optionally) returns a callback that runs with the `ActionResult`.
+To customise the behaviour, you can provide a function that runs immediately before the form is submitted, and (optionally) returns a callback that runs with the `ActionResult`. Note that if you return a callback, the default behavior mentioned above is not triggered. To get it back, call `update`.
 
 ```svelte
 <form
 	method="POST"
-	use:enhance={({ form, data, cancel }) => {
+	use:enhance={({ form, data, action cancel }) => {
```
Comment (geoffrich) [inline documentation/docs/06-form-actions.md:265]:
```suggestion
	use:enhance={({ form, data, action, cancel }) => {
```

File: documentation/docs/06-form-actions.md [hunk 2]
```diff
@@ -278,7 +280,7 @@ You can use these functions to show and hide loading UI, and so on.
 
 #### applyAction
 
-If you provide your own callbacks, you may need to reproduce part of the default `use:enhance` behaviour, such as showing the nearest `+error` boundary. We can do this with `applyAction`:
+If you provide our own callbacks, you may need to reproduce part of the default `use:enhance` behaviour, such as showing the nearest `+error` boundary. Most of the time, calling `update` passed to the callback is enough. If you need more customization you can do so with `applyAction`:
```
Comment (geoffrich) [inline documentation/docs/06-form-actions.md:283]:
```suggestion
If you provide your own callbacks, you may need to reproduce part of the default `use:enhance` behaviour, such as showing the nearest `+error` boundary. Most of the time, calling `update` passed to the callback is enough. If you need more customization you can do so with `applyAction`:
```

File: documentation/docs/06-form-actions.md [hunk 3]
```diff
@@ -307,7 +310,7 @@ If you provide your own callbacks, you may need to reproduce part of the default
 
 The behaviour of `applyAction(result)` depends on `result.type`:
 
-- `success`, `invalid` — sets `$page.status` to `result.status` and updates `form` and `$page.form` to `result.data`
+- `success`, `invalid` — sets `$page.status` to `result.status` and updates `form` and `$page.form` to `result.data` (no exceptions in contrast to `update` from `enhance`)
```
Comment (geoffrich) [inline documentation/docs/06-form-actions.md:313]: Can you expand on this a bit more? I assume you're referring to that `applyAction` will update `form` regardless of the page's location, in contrast to `update`

Comment (dummdidumm) [reply documentation/docs/06-form-actions.md]: I adjusted the defaults explanation in the previous section and enhance the sentence in this - hope this helps

Comment (geoffrich) [reply documentation/docs/06-form-actions.md]: Yeah I think that's better 👍

**Review by geoffrich — APPROVED** [review]

**Pull request merged**